trino/trino.go

```diff
@@ -1253,13 +1263,403 @@ type queryResponse struct {
+//
+// It iterates over `sp.segments`, extracting metadata and handling each segment based on its type:
+// - "inline" segments are decoded and stored directly in `processedSegments`.
+// - "spooled" segments are fetched asynchronously, and their responses are sent to the output channel.
+//
+// The function also maps row offsets to segment indices for ordering.
+//
+// Errors encountered during processing are sent via the returned channel.
+//
+// Returns a receive-only channel (`<-chan httpResponseWithMeta`) that emits processed segment data or errors.
+func (sp *spoolingProtocol) processSegments(
+	processedSegments [][]queryData,
+	mappingSegmentsToArray map[int64]int,
+	dataSize int64,
+) <-chan spooledSegToProccess {
+	ch := make(chan spooledSegToProccess)
+	go func() {
+		defer close(ch)
+		for i, segment := range sp.segments {
+			segMap := segment.(map[string]interface{})
+			metaDataAssertion, _ := segMap["metadata"].(map[string]interface{})
+
+			metaData, err := parseSpoolingMetaData(metaDataAssertion)
+			if err != nil {
+				ch <- spooledSegToProccess{
+					err: err,
+				}
+			}
+
+			dataSize += metaData.RowsCount
+			mappingSegmentsToArray[metaData.RowOffset] = i
+
+			switch segMap["type"] {
+			case "inline":
+				data := segMap["data"].(string)
+				decodedBytes, err := base64.StdEncoding.DecodeString(data)
+				if err != nil {
+					ch <- spooledSegToProccess{
+						err: err,
+					}
+				}
+
+				decodedData, err := decode(decodedBytes, sp.encoder, metaData)
+				if err != nil {
+					ch <- spooledSegToProccess{
+						err: err,
+					}
+				}
+
+				processedSegments[i] = decodedData
+
+			case "spooled":
+				uri, _ := segMap["uri"].(string)
+				ackUri, _ := segMap["ackUri"].(string)
+				headers, _ := segMap["headers"].(map[string]interface{})
+
+				ch <- sp.fetchSegment(uri, ackUri, headers, metaData)
+			}
+		}
+	}()
+
+	return ch
+}
+
+func (sp *spoolingProtocol) fetchSegment(uri, ackUri string, headers map[string]interface{}, metaData spoolingMetaData) spooledSegToProccess {
+	req, err := http.NewRequestWithContext(sp.ctx, "GET", uri, nil)
+	if err != nil {
+		return spooledSegToProccess{err: fmt.Errorf("trino: %w", err)}
+	}
+
+	for k, v := range headers {
+		headerSlice, ok := v.([]interface{})
+
+		if ok {
+			if len(headerSlice) > 1 {
+				return spooledSegToProccess{err: fmt.Errorf("trino: multiple values for header %s", k)}
+			}
+			req.Header.Add(k, headerSlice[0].(string))
+			continue
+		}
+
+		return spooledSegToProccess{err: fmt.Errorf("trino: unsupported header type %T", v)}
+	}
+
+	resp, err := sp.httClient.Do(req)
+	if err != nil {
+		return spooledSegToProccess{err: fmt.Errorf("trino: %w", err)}
+	}
+
+	ackHeaders := make(http.Header)
+	for k, values := range req.Header {
+		for _, v := range values {
+			ackHeaders.Add(k, v)
+		}
+	}
+
+	return spooledSegToProccess{
+		Response:   resp,
+		MetaData:   metaData,
```

Comment:
Why this method just copy the metadata and the ack uri and headers? Maybe it should only return the response?